components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/BaseHistorySuggestionProvider.kt

```diff
@@ -0,0 +1,115 @@
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.browser.icons.IconRequest
+import mozilla.components.concept.awesomebar.AwesomeBar
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.storage.HistoryItem
+import mozilla.components.concept.storage.HistoryMetadata
+import mozilla.components.concept.storage.HistoryStorage
+import mozilla.components.concept.storage.SearchResult
+import mozilla.components.feature.awesomebar.facts.emitHistorySuggestionClickedFact
+import mozilla.components.feature.session.SessionUseCases
+import java.util.UUID
+
+/**
+ * Return 20 history suggestions by default.
+ */
+const val DEFAULT_HISTORY_SUGGESTION_LIMIT = 20
+
+/**
+ * Abstract [AwesomeBar.SuggestionProvider] containing the basic functionality for providing history
+ * results in response to user input changes (tracked by [onInputChanged])
+ *
+ * Child classes are expected to only handle returning a list of [HistoryItem]s in [getResults],
+ * each such child controlling from where and how this results are returned.
+ *
+ * @param historyStorage and instance of the [HistoryStorage] used
+ * to query matching history records.
+ * @param loadUrlUseCase the use case invoked to load the url when the
+ * user clicks on the suggestion.
+ * @param icons optional instance of [BrowserIcons] to load fav icons
+ * for history URLs.
+ * @param engine optional [Engine] instance to call [Engine.speculativeConnect] for the
+ * highest scored suggestion URL.
+ * @param maxNumberOfSuggestions optional parameter to specify the maximum number of returned suggestions,
+ * defaults to [DEFAULT_HISTORY_SUGGESTION_LIMIT]
+ */
+abstract class BaseHistorySuggestionProvider(
+    internal open val loadUrlUseCase: SessionUseCases.LoadUrlUseCase,
+    internal open val icons: BrowserIcons? = null,
+    internal open val engine: Engine? = null,
+    @VisibleForTesting internal open val maxNumberOfSuggestions: Int = DEFAULT_HISTORY_SUGGESTION_LIMIT,
+    override val id: String = UUID.randomUUID().toString()
+) : AwesomeBar.SuggestionProvider {
+
+    /**
+     * Get at most [maxNumberOfSuggestions] history entries for the provided [query].
+     *
+     * @param query Text depending on which to return an appropriate list of history entries.
+     * @return a list containing at most [maxNumberOfSuggestions] [HistoryItem]s.
+     */
+    abstract suspend fun getResults(query: String): List<HistoryItem>
+
+    override val shouldClearSuggestions: Boolean
+        // We do not want the suggestion of this provider to disappear and re-appear when text changes.
+        get() = false
+
+    override suspend fun onInputChanged(text: String): List<AwesomeBar.Suggestion> {
+        if (text.isBlank()) {
+            return emptyList()
+        }
+
+        val historyResults = getResults(text)
+
+        return coroutineScope {
```

Comment:
This doesn't seem right to me. Why do we need a new scope here i.e, why do we launch the speculative connect below an use async/awaitAll. We didn't do that before and are already in a `suspend` method. Why not just:

```kotlin
override suspend fun onInputChanged(text: String): List<AwesomeBar.Suggestion> {
    if (text.isBlank()) {
        return emptyList()
    }

    val historyResults = getResults(text)

    historyResults.firstOrNull()?.url?.let { url -> engine?.speculativeConnect(url) }
    return historyResults.map { it.toAwesomeBarSuggestion() }
}
```

Reply:
Put up an alternative below.

Reply:
The new scope is necessary to be able to call `launch { .. speculativeConnect }`.
The actual speed gains for this method will depend on the host that we're trying to connect to but I see no reason to wait 1 or 10 ms for that call to complete, we can immediately move with mapping the history results.

Reply:
AFAIK, this should not depend on the host as this is not a blocking IO operation?

Reply:
(OK, we double-checked and verified that `speculativeConnect` isn't blocking and we won't get a benefit from launching here. Tested on throttled network and it had no effect on the duration of the call.)